File: api/cues/cuelists.go
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
// SPDX-License-Identifier: MPL-2.0

package cues

const (
	RequestCreateCueList         = "request.cuelist.create"
	RequestUpdateCueListMetadata = "request.cuelist.metadata.update"
	RequestGetCueListMetadata    = "request.cuelist.metadata.get"
	RequestEnumerateCueList      = "request.cuelist.enumerate"
	RequestDeleteCueList         = "request.cuelist.delete"
	RequestMoveCueList           = "request.cuelist.move"

	EventNewCueList    = "event.cuelist.created"
	EventUpdateCueList = "event.cuelist.updated"
	EventDeleteCueList = "event.cuelist.deleted"
)

type CreateCueListInput struct {
	CueListNumber float64 `json:"cueListNumber" msgpack:"cueListNumber" validate:"gte=0"`
}

type CreateCueListOutput struct {
	CueListNumber float64 `json:"cueListNumber" msgpack:"cueListNumber" validate:"gte=0"`
}

type NewCueListEvent struct {
	CueListNumber float64 `json:"cueListNumber" msgpack:"cueListNumber" validate:"gte=0"`
	Label         string  `json:"label" msgpack:"label"`
	ListType      string  `json:"listType" msgpack:"listType"`
}

type UpdateCueListMetadataInput struct {
	CueListNumber float64 `json:"cueListNumber" msgpack:"cueListNumber" validate:"gt=0"`
	Key           string  `json:"key" msgpack:"key" validate:"required"`
	Value         string  `json:"value" msgpack:"value"`
}

type UpdateCueListMetadataOutput struct{}

type UpdateCueListMetadataEvent struct {
	CueListNumber float64 `json:"cueListNumber" msgpack:"cueListNumber" validate:"gt=0"`
	Label         string  `json:"label" msgpack:"label"`
	ListType      string  `json:"listType" msgpack:"listType"`
}

type GetCueListMetadataInput struct {
	CueListNumber float64 `json:"cueListNumber" msgpack:"cueListNumber" validate:"gt=0"`
}

type GetCueListMetadataOutput struct {
	CueListNumber float64 `json:"cueListNumber" msgpack:"cueListNumber"`
	Label         string  `json:"label" msgpack:"label"`
	ListType      string  `json:"listType" msgpack:"listType"`
}

type EnumerateCueListInput struct{}

type EnumerateCueListOutput struct {
	CueLists []GetCueListMetadataOutput `json:"cueLists" msgpack:"cueLists"`
}

type DeleteCueListInput struct {
	CueListNumber float64 `json:"cueListNumber" msgpack:"cueListNumber" validate:"gt=0"`
}

type DeleteCueListOutput struct{}

type MoveCueListInput struct {
	OriginalCueListNumber float64 `json:"originalCueListNumber" msgpack:"originalCueListNumber" validate:"gt=0"`
	NewCueListNumber      float64 `json:"newCueListNumber" msgpack:"newCueListNumber" validate:"gt=0"`
}

type MoveCueListOutput struct {
	OriginalCueListNumber float64 `json:"originalCueListNumber" msgpack:"originalCueListNumber"`
	NewCueListNumber      float64 `json:"newCueListNumber" msgpack:"newCueListNumber"`
}

type DeleteCueListEvent struct {
	CueListNumber float64 `json:"cueListNumber" msgpack:"cueListNumber" validate:"gt=0"`
}

File: frontend/src/lib/stores/cuelists.ts
/**
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * SPDX-License-Identifier: MPL-2.0
 */

import { writable } from 'svelte/store';
import * as Commands from '../../../bindings/gitlab.com/stexxo/dynocue/internal/gui/commands';
import { Events } from '@wailsio/runtime';

export interface CueList {
    cueListNumber: number;
    label: string;
    listType: string;
}

function createCueListStore() {
    const { subscribe, set, update } = writable<CueList[]>([]);

    async function refresh() {
        try {
            const result = await Commands.EnumerateCueList({});
            if (result && result.cueLists) {
                set(result.cueLists);
            }
        } catch (err) {
            console.error('Failed to enumerate cue lists:', err);
        }
    }

    async function updateMetadata(cueListNumber: number, key: string, value: string) {
        try {
            await Commands.UpdateCueListMetadata({ cueListNumber: cueListNumber, key: key, value: value });
            // The store will be updated by the event handler
        } catch (err) {
            console.error(`Failed to update cue list ${cueListNumber} metadata:`, err);
        }
    }

    async function create(cueListNumber: number = 0) {
        try {
            await Commands.CreateCueList({ cueListNumber: cueListNumber });
            // The store will be updated by the event handler
        } catch (err) {
            console.error('Failed to create cue list:', err);
        }
    }

    async function remove(cueListNumber: number) {
        try {
            await Commands.DeleteCueList({ cueListNumber: cueListNumber });
            // The store will be updated by the event handler
        } catch (err) {
            console.error(`Failed to delete cue list ${cueListNumber}:`, err);
        }
    }

    async function move(originalCueListNumber: number, newCueListNumber: number) {
        try {
            await Commands.MoveCueList({ originalCueListNumber: originalCueListNumber, newCueListNumber: newCueListNumber });
            // The store will be updated by the event handlers (delete old, create new)
        } catch (err) {
            console.error(`Failed to move cue list ${originalCueListNumber} to ${newCueListNumber}:`, err);
        }
    }

    // Subscribe to backend events
    Events.On('event.cuelist.created', (event: any) => {
        const { cueListNumber, label, listType } = event.data;
        const result: CueList = {
            cueListNumber: cueListNumber,
            label: label,
            listType: listType
        };
        update(lists => {
            const newList = [...lists, result];
            return newList.sort((a, b) => a.cueListNumber - b.cueListNumber);
        });
    });

    Events.On('event.cuelist.updated', (event: any) => {
        const { cueListNumber, label, listType } = event.data;
        update(lists => lists.map(list => 
            list.cueListNumber === cueListNumber ? { ...list, label: label, listType: listType } : list
        ));
    });

    Events.On('event.cuelist.deleted', (event: any) => {
        const cueListNumber = event.data.cueListNumber;
        update(lists => lists.filter(list => list.cueListNumber !== cueListNumber));
    });

    return {
        subscribe,
        refresh,
        updateMetadata,
        create,
        remove,
        move
    };
}

export const cueLists = createCueListStore();
